ramips: add support for ipTIME A3004NS-dual

HexagonWin opened the pull request:
ipTIME A3004NS-dual is a 2.4/5GHz band router, based on Mediatek MT7621.

Specifications:
- SoC: MT7621 (880MHz)
- RAM: DDR3 256M
- Flash: SPI NOR 16MB
- WiFi:
 - 2.4GHz: MT7602E
 - 5GHz : MT7612E
- Ethernet:
 - 4x LAN
 - 1x WAN
- USB: 1 * USB3.0 port
- UART:
  - 3.3V, TX, RX, GND / 57600 8N1

Installation via web interface:
- 1. Flash Initramfs image using OEM Firmware's web GUI
- 2. Boot into OpenWrt and perform Sysupgrade with sysupgrade image.

Revert to stock firmware:
- 1. Boot into OpenWrt and perform Sysupgrade with OEM Stock Firmware image.

Thanks.

Signed-off-by: Yuchan Seo <hexagonwin@disroot.org>

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:
```diff
@@ -0,0 +1,180 @@
+
+			partition@20000 {
+				label = "config";
+				reg = <0x20000 0x30000>;
```
The reg pair should be `<start length>`, not `<start end>`.

HexagonWin replied:
Thanks, then should I change it to 
`reg = <0x20000 0x10000` ? Thanks.

mans0n replied:
Yes, and please change the other corresponding parts accordingly. (factory, firmware, ...)

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -753,6 +753,17 @@ define Device/iptime_a8004t
 
+define Device/iptime_a3004nd
+  $(Device/dsa-migration)
+  $(Device/uimage-lzma-loader)
```
Just curious, are you sure you need this? What happens if you omit this?

HexagonWin replied:
Hello, OpenWrt does not boot without the $(Device/uimage-lzma-loader).
According to a UART log the router gets lzma error and reboots. 
About $(Device/dsa-migration) : I think this was just copied from iptime-a8004t. I will try removing it and testing.

mans0n replied:
Okay, then other iptime devices might need that remedy. Thanks for the info.
AFAIK you need to keep `$(Device/dsa-migration)` part.

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -753,6 +753,17 @@ define Device/iptime_a8004t
+  UIMAGE_NAME := a3004nd
+  DEVICE_VENDOR := ipTIME
+  DEVICE_MODEL := A3004NS-dual
+  DEVICE_PACKAGES := kmod-usb3 kmod-mt76x2
```
Probably you'll also need `kmod-usb-ledtrig-usbport`.

HexagonWin replied:
Hello, I added the package on the DEVICE_PACKAGES section. Thanks.

mans0n reviewed (commented):
Hi, a few comments below.

mans0n commented:
> Note : The device is very similar to the ipTIME A8004T.

What made you think so? To me it seems it rather resembles WF2881.

HexagonWin commented on target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:
```diff
@@ -0,0 +1,180 @@
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "iptime,a3004nd", "mediatek,mt7621-soc";
```
Hello, for full name, should I then write "iptime,a3004ns-dual", "mediatek,mt7621-soc" then? Thanks.

mans0n replied:
Yes, and please change the other corresponding parts accordingly. (factory, firmware, ...)

HexagonWin commented on target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:
```diff
@@ -0,0 +1,180 @@
+
+/ {
+	compatible = "iptime,a3004nd", "mediatek,mt7621-soc";
+	model = "ipTIME A3004NS-Dual";
```
Thanks, I will edit this part.

HexagonWin commented:
> > Note : The device is very similar to the ipTIME A8004T.
> 
> What made you think so? To me it seems it rather resembles WF2881.

Oh, the patch was based on that device and I could get it to boot with only a tiny change.

HexagonWin commented:
Hello, I fixed those parts and pushed again. Thanks.

mans0n commented:
Please squash your changes into a single commit.
https://openwrt.org/docs/guide-developer/working-with-github-pr#squashing_commits

>>> Note : The device is very similar to the ipTIME A8004T.
>>
>> What made you think so? To me it seems it rather resembles WF2881.
>
> Oh, the patch was based on that device and I could get it to boot with only a tiny change.

That note implies the PCB of A3004NS-dual is almost same with that of A8004T, which is not the case. IMHO that line can be dropped.

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -753,6 +753,17 @@ define Device/iptime_a8004t
 endef
 TARGET_DEVICES += iptime_a8004t
 
+define Device/iptime_a3004nd
```
Please apply proper alphabetic sorting.

HexagonWin commented:
Thanks. I modified based on your comments and pushed again.

mans0n commented:
You still have two commits in this PR.

HexagonWin commented:
Sorry! Fixed.

mans0n commented:
The [From header](https://github.com/openwrt/openwrt/commit/068ae1c381e9067cf0e4dab1ac89a3d0757e944c.patch) should match the Signed-off-by tag, but your patch uses your github username. You can fix this with `git config -e` and `git commit --amend --no-edit --reset-author`.

And `image/Makefile` disappeared from your patch...

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:
```diff
@@ -0,0 +1,180 @@
+
+&state_default {
+	gpio {
+		groups = "wdt", "jtag", "i2c";
```
The device does not use jtag pins, it uses uart3 pins. So this line should be:
`groups = "wdt", "uart3", "i2c";`

HexagonWin replied:
It also has a JTAG port in the board

mans0n replied:
Hmm, you're right. Anyway, if you want to use the pins for jtag purpose, it should not be here, as this node is for gpio pins.

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:
```diff
@@ -0,0 +1,180 @@
+
+			factory: partition@30000 {
+				label = "factory";
+				reg = <0x30000 0x40000>;
```
`reg = <0x30000 0x10000>;`

HexagonWin replied:
Thanks, I'll edit it that way.

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a3004nd.dts:
```diff
@@ -0,0 +1,180 @@
+
+			partition@40000 {
+				label = "firmware";
+				reg = <0x40000 0x1000000>;
```
`reg = <0x40000 0xfc0000>;`

mans0n commented:
And the dts filename should be "mt7621_iptime_a3004ns-dual.dts".

HexagonWin commented:
Hello, I fixed the issues. Thanks

mans0n commented:
The author line still contains your github username. You can check it by `git show --format=fuller`.

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:
```diff
@@ -0,0 +1,180 @@
+
+&state_default {
+	gpio {
+		groups = "wdt", "jtag", "i2c";
```
"uart3" should be included here as the USB LED use the pin.

HexagonWin replied:
I'll add this, Thanks.

HexagonWin commented:
> The author line still contains your github username. You can check it by `git show --format=fuller`.

For me it shows my real name, not alpha2x on both the "Comment" section and commits section. Where does it show the user name? Thanks.

HexagonWin commented:
Fixed it

mans0n commented:
Goto the commit page -> append `.patch` to URL -> Check "From" header in [this page](https://github.com/openwrt/openwrt/commit/1a8e1b7b00138554abe72db3fa405c6ed38c480e.patch)

HexagonWin commented:
Ah.. got it. But how should I fix it?
But the Signed-off-by part shows my name, does the author line still matter..?
Thanks.

mans0n commented:
That's the author line, and it should match SoB in the comment section. (Maybe that's why DCO bot keeps failing?)
`git config user.name "Full Name"`
`git commit --amend --no-edit --reset-author`

HexagonWin commented:
You're right. The error seems to be gone now.
Fixed it.

mans0n reviewed (approved):
`Reviewed-by: Sungbo Eo <mans0n@gorani.run>`
Thanks for your quick feedback!
I'm just a random reviewer here, you'll need to wait for a core dev to get it merged.

CHKDSK88 commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -732,6 +732,17 @@ define Device/iodata_wnpr2600g
 TARGET_DEVICES += iodata_wnpr2600g
 
+define Device/iptime_a3004ns-dual
+  $(Device/dsa-migration)
```
This should be dropped. Your device wasn't supported before DSA migration.

HexagonWin replied:
Thanks, I'll remove it.

CHKDSK88 replied:
My bad, please revert it. Adrian tells me why here: https://github.com/openwrt/openwrt/pull/4416#discussion_r691544161

HexagonWin replied:
@CHKDSK88 Did it..

HexagonWin commented:
Fixed it.

HexagonWin commented:
@CHKDSK88 I tried building the firmware right in this state, and it is working properly.
Is there anything more that I should edit?
Thanks.

CHKDSK88 commented:
> @CHKDSK88 I tried building the firmware right in this state, and it is working properly.
> Is there anything more that I should edit?
> Thanks.

It looks fine for me.

HexagonWin commented:
Hello, may I ask why this isn't being merged...??
I don't see anything more to edit..Thanks.

adschm commented on target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:
```diff
@@ -0,0 +1,180 @@
+		reg = <0x1fc40 0x6>;
+	};
+};
+
```
remove empty line at EOF.

HexagonWin replied:
@adschm Hello, I fixed it. Thanks.

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -732,6 +732,17 @@ define Device/iodata_wnpr2600g
+  UIMAGE_NAME := a3004nd
+  DEVICE_VENDOR := ipTIME
+  DEVICE_MODEL := A3004NS-dual
+  DEVICE_PACKAGES := kmod-usb3 kmod-mt76x2
```
Please add `kmod-usb-ledtrig-usbport` here.

HexagonWin replied:
Odd, it disappeared.. Thanks.

HexagonWin replied:
I don't know what I did but the name just reverted.. I'll just fix them all and re-post. Sorry!

HexagonWin replied:
Done

HexagonWin commented:
Hello, if there aren't more issues, can this please be merged? Thanks.

HexagonWin commented:
@adschm I'm sorry but can you please check if there are more issues with the PR? Thanks.

HexagonWin committed:
ramips: add support for ipTIME A3004NS-dual

ipTIME A3004NS-dual is a 2.4/5GHz band router, based on Mediatek MT7621.

Specifications:
- SoC: MT7621 (880MHz)
- RAM: DDR3 256M
- Flash: SPI NOR 16MB
- WiFi:
 - 2.4GHz: MT7602E
 - 5GHz : MT7612E
- Ethernet:
 - 4x LAN
 - 1x WAN
- USB: 1 * USB3.0 port
- UART:
  - 3.3V, TX, RX, GND / 57600 8N1

Installation via web interface:
- 1. Flash Initramfs image using OEM Firmware's web GUI
- 2. Boot into OpenWrt and perform Sysupgrade with sysupgrade image.

Revert to stock firmware:
- 1. Boot into OpenWrt and perform Sysupgrade with OEM Stock Firmware image.

Signed-off-by: Yuchan Seo <hexagonwin@disroot.org>

adschm commented on target/linux/ramips/dts/mt7621_iptime_a3004ns-dual.dts:
```diff
@@ -0,0 +1,179 @@
+
+&state_default {
+	gpio {
+		groups = "wdt", "jtag", "i2c", "uart3";
```
I don't see a GPIO for jtag?

HexagonWin replied:
Hello, I'm sorry but I don't know much about this, is another jtag node needed?
The jtag pins are available on board. The similar for the supported ipTIME A8004T, it's dts ( https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621_iptime_a8004t.dts ) also seems to not have a jtag node. Thanks.

mans0n replied:
If you're going to use the pins for gpio purpose, then leave it as it is. (though I do not prefer it personally)
For jtag purpose, you might need to add a node like this:
```
	jtag {
		groups = "jtag";
		function = "jtag";
	};
```
...or skip it if the bootloader already enabled it.
AFAIK A8004T does not expose jtag pins so its setup is right.
I don't care much about it here, but I believe it's not a good practice to add a feature that has not been tested. Have you tested the gpio function with the jtag pins? Or did you add the node just because the pins were there?

HexagonWin replied:
@mans0n: I just left it since the pins were available.
Unfortunately I do not have any kinds of testing devices, especially none for jtag.

adschm replied:
I would have simply removed the jtag group?

HexagonWin replied:
Alright, I'll just remove it then.
Should I just remove the thing by changing
```
groups = "wdt", "jtag", "i2c", "uart3";
```
to?
```
groups = "wdt", "i2c", "uart3";
```

mans0n replied:
Precisely.

HexagonWin replied:
Done.

HexagonWin commented:
@adschm: I successfully checked the patch working on latest master again. Could you please check again if there are more issues with the PR? Thanks.

mans0n commented:
Thanks! Merged as c33eb0372e7c.
